File: frontend/src/features/notes/note.api.ts
import { api, query } from '../../lib/api/client'
import type { Note, NoteImage, NoteInput, Page } from './note.types'
export const listNotes = (params: Record<string, string | number | undefined> = {}) =>
  api<Page<Note>>(`/v1/notes${query(params)}`)
export const getNote = (id: string) => api<Note>(`/v1/notes/${id}`)
export const createNote = (input: NoteInput) =>
  api<Note>('/v1/notes', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  })
export const updateNote = (id: string, input: Partial<NoteInput>) =>
  api<Note>(`/v1/notes/${id}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  })
export const deleteNote = (id: string) => api<void>(`/v1/notes/${id}`, { method: 'DELETE' })
export const uploadImage = (noteId: string, file: File) => {
  const form = new FormData()
  form.set('file', file)
  return api<{ image: NoteImage; token: string }>(`/v1/notes/${noteId}/images`, {
    method: 'POST',
    body: form,
  })
}
export const imageAccess = (id: string) =>
  api<{ url: string; expires_in: number }>(`/v1/note-images/${id}/access`)
export const updateImageAlt = (id: string, alt_text: string | null) =>
  api<NoteImage>(`/v1/note-images/${id}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ alt_text }),
  })
export const deleteImage = (id: string) => api<void>(`/v1/note-images/${id}`, { method: 'DELETE' })

File: android/app/src/main/java/com/tasker/android/remote/api/NoteApi.kt
package com.tasker.android.remote.api

import com.tasker.android.remote.dto.ItemsResponse
import com.tasker.android.remote.dto.NoteCreateRequest
import com.tasker.android.remote.dto.NoteDto
import com.tasker.android.remote.dto.NoteImageUploadResponse
import com.tasker.android.remote.dto.NoteUpdateRequest
import com.tasker.android.remote.dto.SignedUrlResponse
import okhttp3.MultipartBody
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.Multipart
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Part
import retrofit2.http.Path
import retrofit2.http.Query

interface NoteApi {

    @GET("v1/notes")
    suspend fun listNotes(
        @Query("q") query: String? = null,
        @Query("limit") limit: Int = 1000,
    ): ItemsResponse<NoteDto>

    @POST("v1/notes")
    suspend fun createNote(@Body body: NoteCreateRequest): NoteDto

    @GET("v1/notes/{id}")
    suspend fun getNote(@Path("id") id: String): NoteDto

    @PATCH("v1/notes/{id}")
    suspend fun updateNote(@Path("id") id: String, @Body body: NoteUpdateRequest): NoteDto

    @DELETE("v1/notes/{id}")
    suspend fun deleteNote(@Path("id") id: String): Response<Unit>

    @PUT("v1/notes/{noteId}/tasks/{taskId}")
    suspend fun linkTask(@Path("noteId") noteId: String, @Path("taskId") taskId: String): Response<Unit>

    @DELETE("v1/notes/{noteId}/tasks/{taskId}")
    suspend fun unlinkTask(@Path("noteId") noteId: String, @Path("taskId") taskId: String): Response<Unit>

    @Multipart
    @POST("v1/notes/{noteId}/images")
    suspend fun uploadImage(
        @Path("noteId") noteId: String,
        @Part file: MultipartBody.Part,
    ): NoteImageUploadResponse

    @DELETE("v1/note-images/{imageId}")
    suspend fun deleteImage(@Path("imageId") imageId: String): Response<Unit>

    @GET("v1/note-images/{imageId}/access")
    suspend fun getSignedUrl(@Path("imageId") imageId: String): SignedUrlResponse
}
